Task01/Var1 - Генерация оптимизированного графа по txt

ZloiZebroo opened the pull request:
Сделал генерацию оптимизированного графа по тхт файлу и сохранение его в json

ZloiZebroo committed:
- убрал папку .idea через gitignore

ZloiZebroo committed:
Merge branch 'task01' into main

ZloiZebroo committed:
- поправил readme

ZloiZebroo committed:
- поправил readme

ZloiZebroo committed:
Merge branch 'task01' into main

ZloiZebroo committed:
- добавил класс "лабиринт"
- сделал читалку тхт файла с лабиринтом

ZloiZebroo committed:
Сделал генерацию графа

- сделал генерацию  оптимизированного графа
- сделал сохранение графа в json
- сделал чтение графа из json
- сделал генерацию матрицы смежнонсти

ZloiZebroo committed:
Json стал красивее

- сделал json красивым

ZloiZebroo committed:
Merge branch 'dz_01' into main

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+        self.adj_matrix = None
+
+    def load_from_txt(self, txt_path):
+        file = open(txt_path, "rt")
```
Файлы (ресурсы) лучше использовать с with - это гарантирует, что они будут закрыты

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+        print(self.adj)
+        self.adj_matrix = adjacency_matrix(self.nodes, self.adj)
+
+    def to_json(self):
```
Немного забавно, что при чтении лабиринты вы передаете имя входного файла, а при записи - нет. Даже в "костыльном коде" методы и функции лучше оформлять так, чтобы передавать в них аргументы, а не использовать хардкоженные переменные

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+        self.is_node = False
+
+
+class Maze:
```
Лучше (не знаю почему, но мне кажется PyCharm так говорит) явно наследоваться от object

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+    maze_2 = 'maze_02.txt'
+    maze_json = 'maze_graph.json'
+
+    # Читаем лабиринт из тхт файла
```
Такие комментарии никогда не пишите, это прекрасно видно уже из вашего кода, naming достаточно хороший. Комментарии такого рода только усложняют чтение кода. Он имел бы смысл только в том случае, если у вас вдруг в `load_from_txt` оказалось бы что-то еще (но естественно так лучше не делать :))

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+    def __init__(self):
+        self.matrix = None
+        self.nodes = None
+        self.adj = None
```
Кстати, вопрос к нейменгу и комментариям, вот тут они были бы полезны. В чем отличие `self.adj` от `self.adj_matrix`?

ZloiZebroo replied:
`self.adj` - связи между графами
`self.adj_matrix` - матрица смежности

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+    maze_json = 'maze_graph.json'
+
+    # Читаем лабиринт из тхт файла
+    maze = Maze()
```
Касательно класса Maze - его архитектурная надобность здесь стремится к нулю. Единственное, что он агрегирует для совместной обработки - это self.nodes и self.adj
В то время когда вы строите матрицу связанности поля класса никак не использованы, все равно просто передается в отдельную функцию. Если уже делать классы - я бы в первую очередь озаботился бы именно классом для матрицы связанности, так как это наиболее нагруженная сущность во всем коде.

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+import json
```
Структура директорий неправильна - первое ДЗ должно быть в папке task01

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+
+    # Читаем лабиринт из тхт файла
+    maze = Maze()
+    maze.load_from_txt(maze_2)
```
Действительно в методе оказалась не только загрузка, но и парсинг лабиринта. В целом наверное стоило бы разделить эти методы. Вынести все, что относится к чтению файла во вне. Кроме этого в данном случае лучше бы как раз подошли classmethod или staticmethod (фабричный метод). То есть, метод который создает объект из какой-то заданной сущности.

ZloiZebroo replied:
Я посчитал, что целевой сущностью в данном случае является сам граф. Методы `load_from_txt`  и `load_from_json` дают на выходе один и тот же целевой граф, который мы хотели получить, а парсинг тхт файла - побочный эффект при чтении данных в таком формате.

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+
+    last_y = len(data) - 1
+    last_x = len(data[0]) - 1
+    print(f'Последний индекс по Х: {last_x}')
```
У вас логирование где-то на русском, где-то на английском (и комментарии тоже). Лучше уже придерживаться одного языка во всей программе. Общепринятым конечно является английский

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+    print(f'Последний индекс по Х: {last_x}')
+    print(f'Последний индекс по Y: {last_y}')
+
+    for y in range(len(data)):
```
#itertools
Я бы заюзал здесь rangend, который писал на семинарах

#Нейминг
Когда работает с матрицами очень не советую именовать индексы x, y. Потому что, как вы могли заметить они идут в другом порядка по отношению к обычной геометрии. Кроме этого ось Y еще и вниз направлена. Поэтому лучше использовать всегда терминологию row и col (r, c).

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+
+
+# Возможные пути по которым можно пойти от точки
+def find_possible_ways(data, point):
```
Вот здесь начинает прослеживаться "ошибка" дизайна классов. У вас в нескольких функция появляется одна и та же операция, получить last_x, last_y. Если уж погнали делать классы - обернули data - в класс Matrix, у которого уже сделали бы метода (или так как матрица не меняется) предвычислили бы кол-во ROW и COL.

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+def find_possible_ways(data, point):
+    possible_ways = list()
+    x, y = point
+    last_y = len(data) - 1
```
Здесь и выше и ниже: не очень хорошо использовать last.... в контексте матриц. Обычно используется размер или количество row, col. А когда нужен ласт, то делают  - 1

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+# Возможные пути по которым можно пойти от точки
+def find_possible_ways(data, point):
+    possible_ways = list()
+    x, y = point
```
Для чего вы передаете в функцию tuple? По факту только нагрузили ваш код лишними действиями и тут даже вопрос не столько к скорости сколько к readability. Зачем-то собираете xy в tuple, чтобы следующей операцией его разделить.

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+                return nodes, ways, previous_point
+
+
+# Возможные пути по которым можно пойти от точки
```
Вопрос к неймингу: конечно это навязывание, но я бы назвал это соседями, так как один шаг - сложно назвать путем. Либо связями. Пути - очень запутывают.

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+
+    for y in range(len(data)):
+        for x in range(len(data[0])):
+            nodes, ways, is_node, point = get_node(x, y, data, nodes)
```
#Дизайн_классов
Вы сделали класс точки Point, но используете его очень ограниченно и получается, что из функции возвращаете кучу переменных через тупл. Одна из которых - это bool (это вообще ужасно). Почему не назвать класс Point, классом Node и не запихать туда все эти переменные?
Ну is_node поменять на is_simple например? А ways я бы заменил на links - это же связи с соседями.

#Неправильно_использован_tuple
Возвращать из функцию кучу переменных через тупл - это прям кричащее место, где нужен просто правильный дизайн классов. Это вносит жуткую путаницу в код и скорее всего через пару недель вы сами забудете, что эта функция возвращала. Обращайте на такие места и постарайтесь больше так не делать в коде (особенно когда, что-то будет писать в прод). Более двух переменных не возвращайте через тупл.

#Возврат_входного_аргумента
Для чего в данном случае вы возвращаете nodes? Он является аргументом функции и изменяется в ней. Вы нигде не используете эту функцию как было у меня в примере, чтобы не инициализировать результат работы. На мой взгляд это просто не нужно действие

PS: я видел, что конкретно эта функция find_nodes не юзается - это кстати тоже плохо, когда отдаете код в прод, такие вещи нужно убирать, мертвый код усложняет читаемость.

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+
+
+def convert_to_graph(data):
+
```
nodes я бы объявил прямо здесь и уже везде передавал как словарик со нодами графа

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+    nodes, ways, previous_point = find_start_node(data)
+
+    # снарядили экспедицию
+    wte = WaysToExplore(ways, previous_point)
```
Вот здесь вы как раз оборачиваете точку еще в один класс у которого прописаны соседи. Почему это все:
1. Не упихать в саму точку?
2. Сразу не вернуть из get_node?

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+
+    # Если больше некуда идти, то выходим
+    # а если есть, то продолжаем
+    if len(dc.next_explore) > 0 and dc.point.id < 10:
```
#Комментарии
Что за сравнение с 10? Почему нету комментария по этому поводу

ZloiZebroo replied:
Осталось от дебага, уже поправил. Чтобы функция не улетала в бесконечный цикл, стоит ограничение на 10 id узлов.

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+    return dc.nodes, dc.adj
+
+
+def adjacency_matrix(nodes, adj):
```
Если я все правильно понял, то здесь эффектиность - отрицательная :)

```python
def adjacency_matrix(nodes, adj):
    size = len(adj)
    matrix = [[None] * size for _ in range(size)]
    for node, links in adj.items():
        row = int(node)
        for other, length in links.items():
            col = int(other)
            matrix[row][col] = length
    return matrix
```
разве так не проще?

И зачем вообще словарь nodes? Не очень понял

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+                    string.append(1)
+                else:
+                    string.append(0)
+            except KeyError:
```
Если есть возможность передавать полезную информация не через исключения - лучше тогда их не использовать. Исключения всегда медленнее обычных сравнений. То есть в данном случае лучше использовать get()

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+
+
+# Чтобы удобнее было таскать данные по функциям
+class DataCase:
```
Что касается этого класса.... на мой взгляд - он все очень усложняет. Проблема в том, что когда начинаешь разбираться с функцией `recursive_explorer` возникает оооочень большая связанность. Так как получается так, что эти данные все фактически передаются как изменяемые. Ну то есть условно вы "передали по указателю". Получается в каждый из них может в любой момент измениться и что сейчас с ними происходит ... сложно понять.

Первое чтобы я переделал при рефакторинге - это отказался бы от этого класса и переделал бы на просто переменные в функции recursive_explorer. Дальше уже по взаимосвязям смотрел, что объединить в классы и как

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+                # увеличим длинну пути
+                dc.way_len += 1
+
+                # Если больше нет путей потыкаться, ливаем с катки
```
Смотри так и бан пропишут на 30 минут

xthebat commented on ZloiZebroo/Dz01/maze.py:
```diff
@@ -0,0 +1,332 @@
+                point_id = len(nodes)
+                nodes[point_id] = [x, y]
+            else:
+                point_id = list(nodes.values()).index([x, y])
```
Вот тут бы пригодился инвертированный словарь. Это кстати в сторону того, что не очень хорошо задизайнен nodes во время преобразования.

Он отображает номер узла в индексы клетки. Мне кажется, что нужно как раз наоборот, так операция получить номер клетки по имени узла - судя по коду вообще очень редкая. 

А пара индексов - это уникальное значение. Кроме того, когда дампишь в json все равно численные ключи преобразовал в str

ZloiZebroo replied:
Да, с инвертированным словарём было бы удобнее, но зачем таскать с собой инвертированную копию словаря, если нужную инфу можно вытащить из уже имеющегося. Если текущий словарь заменить на инвертированный, то будет не такой красивый выходной json. Там координаты узлов и пути связанны как раз по номеру узла.

Численные ключи преобразовываются в строки из-за особенности работы `json.dumps() `.  Жертвуя численными ключами, получаем красивый отсортированный json файл с отступами.

xthebat replied:
Выгрузка в файл операция достаточно редкая, а вот поиск элементов у вас выполняется почти каждую рекурсию, если не чаще. Поиск по значениям в словаре это крайне не эффективно по сравнению поиска по ключам. Поэтому лучше использовать в алгоритме инвертированный словарь, а при выгрузке только для сохранения инвертировать его временно в памяти.

ZloiZebroo replied:
TypeError: unhashable type: 'list'

ZloiZebroo replied:
решаемо

xthebat replied:
Используйте tuple - он не изменяем и хешируем

TLTR: Правильно, потому что в python list - нехешируемый (я об это говорил). То есть от объектов типа list посчитать хешь нельзя. А этот самый хеш нужен, чтобы объект был ключом словаря (или мог быть добавлен в множество). Хеш позволяет значительно сократить время поиска объекта в коллекции. У словаря нету хеша - так как словарь изменяемый тип. Представьте, что если бы список был хешируем, то вы могли бы добавить в словарь список, а потом по ссылке его изменить, так чтобы он стал равен другому объекту. У вас в итоге получиться словарь с двумя одинаковыми ключами - это порушит все.

ZloiZebroo replied:
я уже list просто перевёл в string и всё завелось)

xthebat replied:
Ох, я понимаю, что это учебная задача :) и особо прироста в скорости не будет. Но вы когда лист к строке приводите, чтобы как ключ использовать - это прям жуткий костыль и опять трата процессора. Почему сразу туплы не делать? Это как раз идеальный случай где они нужны. У вас координаты и никогда не меняются

ZloiZebroo replied:
Вы говорили, что туплы плохо сочетаются с json. Их всё равно потом придётся в список перегонять.

ZloiZebroo replied:
Перевёл вектор координаты в туплы и сделал в конце конвертацию в массив

xthebat replied:
Да, плохо сочетаются с json - но перегнать в файл - это не очень часто (и текущей задаче и в целом, когда сохраняем что-то в файл, скорости ждать не приходится). А вот обращение в памяти к координатам во время рекурсии постоянно, поэтому тут стоит экономить процессор. А перед сохранением уже конвернтуть во что удобнее будет сохрнять

xthebat reviewed (changes requested):
"+" Кода написан в соответствии с PEP8, нет ни одной ругани от PC
"+" Структура функций неплохая
"+" Нейминг в целом норм

"-" Дизайн классов - жесть
"-" Дизайн возвращаемых аргументов - жесть

"-" Не работает на лабиринте:
```
11
11
```
"-" Не правильно работает на лабиринте - должно быть 9 узлов, в результате 5 узлов
```
111
111
111
```

Итого:
Необходима правильна работа на приведенных лабиринтах и юнит-тесты. Рефакторинг по желанию и после исправления не правильной работы.

И да, дальше по заданию:
- Поиск кратчайшего пути
- Опционально: удаление стенки

ZloiZebroo commented:
Что такое дизайн классов?

xthebat commented:
Это то как вы спроектировали вашу структуру/архитектуру классов. Какой класс с кем связан, для чего он и какие методы у него и как они должны быть использованы.

xthebat commented:
И да насчет лабиринта 3х3, я видимо ошибся, всё правильно - должно быть 6 узлов

ZloiZebroo committed:
- поправил некоторые замечания

ZloiZebroo committed:
Поправил

- поправил все предыдущие замечания
- сделал поиск кратчайшего пути диикстрой
- сделал поиск кратчайшего пути с удалением стенки

ZloiZebroo committed:
не загрузились в прошлый комит

- Догружаю отвалившееся

ZloiZebroo committed:
Merge pull request #1 from ZloiZebroo/dz_01

Dz 01

ZloiZebroo commented:
Поправил предыдущие замечания и сделал доработки:

- Добавил юнит тесты
- Добавил поиск кратчайшего пути Дийкстрой
- Использовал Rengend
- Удалил класс Maze
- Теперь работает на лабиринте 2х2 и 3х3
- Добавил поиск кратчайшего пути с удалением стенки
- Наследовался от object в классах
- Новый дизайн классов
- Новый преобразователь в матрицу смежности
- Убрал не используемую функцию Find_nodes
- Переименовал пути в соседей
- Логирование на английском
- Передача пути сохранения файла через аргумент
- Поправил структуру папок
- С файлами использовал with

xthebat commented on ZloiZebroo/task01/maze.py:
```diff
@@ -22,215 +22,194 @@
 
 
 # получить узел, если это узел
```
В прошлый раз не обратил внимание: документация в питоне пишется внутри функции в тройных ковычках, например:

```python
def say_meow(arg):
    """this function just say meow"""
    print("meow")
```

xthebat commented on ZloiZebroo/task01/maze.py:
```diff
@@ -22,215 +22,194 @@
 
 # получить узел, если это узел
-def get_node(x, y, data, nodes):
+def get_node(point, matrix, nodes):
```
С этой функцией очень страшно - она называется get_node (получить узел), но внутри меняет список узлов. Это не очень хорошо. То есть нужно стараться писать код так, чтобы не перегружать различной функциональностью одну функцию. Кроме этого - эта функция зовется в find_start_node и тоже начинает модифицировать список узлов. Я хотел распутать этот узел - но после модификации узлов стало совсем страшно, частично можно посмотреть здесь:
https://github.com/xthebat/iu4-2k21-mlni/commit/7abd374f6f6c645afb42de5d952e41bae880340c

ZloiZebroo replied:
`get_node` автоматом добавляет узел в список узлов, чтобы потом не надо было это делать вручную. Сразу его и определяет и кладёт в "корзинку", а остальным функциям остаётся только её пинать в нужный момент.  

`find_start_node` находит первый попавшийся узел, чтобы начать с него обход лабиринта.

xthebat commented on ZloiZebroo/task01/maze.py:
```diff
@@ -22,215 +22,194 @@
+
+def find_start_node(nodes, matrix):
+
+    for r in range(len(matrix.data)):
```
У вас же есть columns, rows

xthebat commented on ZloiZebroo/task01/maze.py:
```diff
@@ -22,215 +22,194 @@
                 # Новые Х и У, так как это не узел, путь у него 1
-                x = wte.ways_to_explore[0][0]
-                y = wte.ways_to_explore[0][1]
+                # x = wte.ways_to_explore[0][0]
```
мертвый код нужно удалять, возможно смотрю не последний коммит)

xthebat commented on ZloiZebroo/task01/maze.py:
```diff
@@ -241,39 +220,46 @@ def adjacency_matrix2(nodes, adj):
+class Graph(object):
 
     def __init__(self):
         self.matrix = None
```
Во тут я потерялся, последний коммит в main в вашем репозитории другой код по сравнению с этим коммитом. Причем если этот коммит не последний, то в нем Graph реализован лучше. 

То что я увидел в последнем коммите, вы создаете просто Graph, а потом ему накидываете на поля уже значения, которые получены в какой-то другой сторонней функции. Это фактически рушит принцип ООП. Смысл же в том, что у нас есть данные методы их обработки и это в одном месте. 

То есть, если у вас ситуация, что для создания объекта нужны какие-то данные, то эти данные должны идти в конструктор, а не присваиваться где-то потом. Если для вычисления чего-либо нужные какие-то данные, то скорее всего эти функции это методы класса в котором данные содержаться. К какому классу отнести методы, если вроде как нужно для обоих классов - смотрим по связанности с какими данными методы связаны больше к тем классам и относим

ZloiZebroo replied:
Вот тут актуальная версия. Вроде бы в этом реквесте она же должна быть, потому что он подтягивает её из main

https://github.com/ZloiZebroo/iu4-2k21-mlni/commit/033089f8e60c4d51c1e7ec97eb8878103e11e5db

xthebat replied:
Входной файл:
```
11110111
00000001
11110111
11110111
```

код теста:
```python
    def test_maze_08(self):
        data = graph.load_from_txt('../maze_08.txt')
        ans1, ans2 = graph.find_start_node(dict(), data.matrix)
        print(data)
        print(ans1)
        print(ans2)
        print("meow")
```

Смотрю в дебаге:
![image](https://user-images.githubusercontent.com/2856140/112596244-9b42b200-8e1c-11eb-92ce-3caf0e38e299.png)

Вопросы и комментарии:
- Ноды до поломки - ок
- Матрицы до поломки - ок
- Стенки для поломки - судя по всем все возможные, можно ломать только одну.
- Я так понимаю алгоритм пока не строит пути? Если строит, то как их получить?

ZloiZebroo replied:
Путь печатается в консоль

ZloiZebroo replied:
Я так понял, в этом лабиринте проверяется случай, когда без поломки стенки невозможно добраться до выхода. Я такой вариант исключил из алгоритма и предполагал, что поломка стенки будет позволять сокращать уже имеющийся путь. Как раз мёртвая функция  `find_nodes` из предыдущих комитов несла в себе идею просветки всего лабиринта насквозь и составления карты узлов заранее, но я посчитал, что такого случая не будет, потому что кому надо стучаться в изолированные узлы, если в них всё равно не попасть (без поломки стенки).

xthebat replied:
Тогда не нашел, где есть тест на сам путь. То есть печатается - окей, но нужно бы тест. В этом тесте тогда сразу можно поиграться.
В итоге:
- поиску пути как я понял есть? Если так, то нужно вынести в тест
- насколько помню говорил, что в исходном лабиринте пути может не быть :) Ну окей, если забыл, будем считать, что путь должен быть и в исходном лабиринте. 
- Если поиск пути с поломкой стенки есть, вынеси пожалуйста это в отдельный тест, где проверяется найденный путь.

xthebat reviewed (commented):
Тесты я посмотрел, попробую позадавать разные лабиринты, если на них пройдет, то тогда принято. Но с классами тяжеловато :) 
Вот тут основные принципы расписаны, без примеров, к сожалению:
https://tproger.ru/translations/oop-principles-cheatsheet/

Ну наверное вот тут неплохо про классы (ООП) с примерами, но многобукв
https://realpython.com/inheritance-composition-python/

ZloiZebroo committed:
Поправил замечания

- Тест для кратчайшего пути
- Тест для пути с пробиванием стенки
- Пофиксил неточность, когда все разбитые стенки становились узлами

ZloiZebroo committed:
Merge pull request #2 from ZloiZebroo/dz_01

Поправил замечания

ZloiZebroo committed:
Небольшие корректировки

- minor fix

ZloiZebroo committed:
Merge branch 'dz_01' into main

ZloiZebroo commented:
Как этим пользоваться:

`load_from_txt` загружает оптимизированный граф 
`shortest_way` позволяет найти кратчайший путь между двумя узлами этого графа

`shortest_way(0, 3, graph, brake_wall=True)`
`0` - id начального узла
`3` - id конечного узда
`graph` - загруженный с помощью `load_from_txt` граф
` brake_wall=True` - параметр, позволяющий выбрать режим поиска кратчайшего пути. В положении True, будет производится поиск с возможностью удаления одной стенки. При этом, очерёдность id узлов не поменяется, а новообразовавшиеся узлы добавятся в конец словаря `nodes`.

ZloiZebroo committed:
Улучшил поиск пути

- теперь ищит путь, даже если конечный узел  изначально находится за стенкой
- в тестах поменялся только порядок  значений expected

ZloiZebroo committed:
Merge branch 'dz_01' into main

ZloiZebroo committed:
Удалил тест поиска стартовог узла

ZloiZebroo committed:
Merge branch 'main' into dz_01

ZloiZebroo commented:
Что нового:

- Работает на графе, в котором изначально невозможно добраться до выхода без пробития стенки
- Появились тесты для поиска пути
- Работает на пустом графе
- В комментарии выше инструкция как с этим играться 
- Кратчайший путь и его длинна теперь хранятся в объекте `Graph`, а не только в консоли

ZloiZebroo committed:
Ускорил работу алгоритма в 2 раза

- избавился от deepcopy

ZloiZebroo committed:
Merge branch 'dz_01' into main

Commit:
Merge branch 'main' into main
